src/jl_uv.c

```diff
@@ -693,8 +694,59 @@ JL_DLLEXPORT void jl_safe_printf(const char *fmt, ...)
     va_end(args);
 
     buf[999] = '\0';
-    if (write(STDERR_FILENO, buf, strlen(buf)) < 0) {
-        // nothing we can do; ignore the failure
+    // Our telemetry on SPCS expects a JSON object per line
+    // The following lines prepare the timestamp string and the JSON object
+    struct timeval tv;
+    struct tm* tm_info;
+    char timestamp_buffer[50];
+    // Get current time
+    gettimeofday(&tv, NULL);
+    tm_info = gmtime(&tv.tv_sec);
+    // Format time
+    int offset = strftime(timestamp_buffer, 25, "%Y-%m-%dT%H:%M:%S", tm_info);
+    // Append milliseconds
+    snprintf(timestamp_buffer + offset, 25, ".%03d", tv.tv_usec / 1000);
+
+    const char *json_preamble_p1 = "\n{\"level\":\"Error\", \"timestamp\":\"";
+    const char *json_preamble_p2 = "\", \"message\": \"";
+    const char *json_postamble = "\"}\n";
+    if (jl_inside_signal_handler() && jl_sig_fd != 0) {
+        // Ignore write failures because there is nothing we can do
+        write(jl_sig_fd, json_preamble_p1, strlen(json_preamble_p1));
+        write(jl_sig_fd, timestamp_buffer, strlen(timestamp_buffer));
+        write(jl_sig_fd, json_preamble_p2, strlen(json_preamble_p2));
+        // JSON escape the input string
+        for(size_t i = 0; i < strlen(buf); i += 1) {
+            switch (buf[i]) {
+                case '"':
+                    write(jl_sig_fd, "\\\"", 2);
+                    break;
+                case '\b':
+                    write(jl_sig_fd, "\\b", 2);
+                    break;
+                case '\n':
+                    write(jl_sig_fd, "\\n", 2);
+                    break;
+                case '\r':
+                    write(jl_sig_fd, "\\r", 2);
+                    break;
+                case '\t':
+                    write(jl_sig_fd, "\\t", 2);
+                    break;
+                case '\\':
+                    write(jl_sig_fd, "\\\\", 2);
+                    break;
+                default:
+                    write(jl_sig_fd, buf + i, 1);
+            }
+        }
+        write(jl_sig_fd, json_postamble, strlen(json_postamble));
+        fdatasync(jl_sig_fd);
+    }
+    else {
+        if (write(STDERR_FILENO, buf, strlen(buf)) < 0) {
+            // nothing we can do; ignore the failure
+        }
```

Comment:
That's the behavior we want to avoid I believe: we're literally using this flag to ensure logs **don't** go to some "unsafe" file descriptor such as stdout or stderr.

Reply:
No, the intent is not to remove output from `stderr`. See the Slack thread. I think we should leave the existing behavior unchanged and just add the write of the JSON-ified message to the signal log file.

Reply:
SGTM then. Thanks for the clarification.

Reply:
@adnan-alhomssi, @li1: please confirm Kiran's interpretation is the functionality you folks want.

Reply:
"stderr" goes to a log file that we can get on-demand by asking the customer to pass its content to us.
We want to treat the stacktrace as "safe" by writing it to the dedicated log file for safe logs.

In addition to writing to the safe log if it is set, we can also always write to stderr (second copy).

Does that answer you?

Reply:
Yes, thanks.